File: sdks/typescript/src/context.ts
/**
 * Request-scoped metadata propagated via Node's AsyncLocalStorage.
 *
 * This is the async-propagating equivalent of Python's ContextVar: values set
 * for the duration of `requestContext(ctx, fn)` are visible to every nested
 * LLM call inside `fn` — including across `await` boundaries — while staying
 * isolated between concurrent async flows.
 */

import { AsyncLocalStorage } from "node:async_hooks";

export interface RequestContext {
  customerId?: string;
  userId?: string;
  traceId?: string;
  /** Any extra key/value pairs become tags (or known fields if recognized). */
  [key: string]: unknown;
}

const storage = new AsyncLocalStorage<Record<string, unknown>>();

/** Return the current request-context object (empty if none set). */
export function currentContext(): Record<string, unknown> {
  return { ...(storage.getStore() ?? {}) };
}

function clean(d: Record<string, unknown>): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(d)) {
    if (v !== undefined && v !== null) out[k] = v;
  }
  return out;
}

/**
 * Run `fn` with `ctx` attached as request-scoped metadata. Works for sync and
 * async functions: the return value (including a Promise) is passed straight
 * through. Nesting accumulates — an inner context merges over the outer one.
 *
 * @example
 *   await requestContext({ customerId: "c1", traceId: "t1" }, async () => {
 *     await client.chat.completions.create({ ... });
 *   });
 */
export function requestContext<T>(ctx: RequestContext, fn: () => T): T {
  const merged = {
    ...(storage.getStore() ?? {}),
    ...clean(ctx as Record<string, unknown>),
  };
  return storage.run(merged, fn);
}

File: sdks/go/context.go
package costobs

import "context"

// Metadata is request-scoped attribution attached to observed calls.
// Unknown attribution should go in Tags.
type Metadata struct {
	Environment   string
	Team          string
	Service       string
	CustomerID    string
	UserID        string
	TraceID       string
	Feature       string
	PromptKey     string
	PromptVersion string
	Tags          map[string]string
}

type ctxKey struct{}

// WithMetadata returns a context carrying metadata for calls made beneath it.
// Nested calls merge: inner non-zero fields override outer ones, tags union.
func WithMetadata(ctx context.Context, m Metadata) context.Context {
	if existing, ok := ctx.Value(ctxKey{}).(Metadata); ok {
		m = mergeMetadata(existing, m)
	}
	return context.WithValue(ctx, ctxKey{}, m)
}

// MetadataFrom extracts the active metadata (zero value if none).
func MetadataFrom(ctx context.Context) Metadata {
	if m, ok := ctx.Value(ctxKey{}).(Metadata); ok {
		return m
	}
	return Metadata{}
}

func mergeMetadata(base, over Metadata) Metadata {
	pick := func(b, o string) string {
		if o != "" {
			return o
		}
		return b
	}
	out := Metadata{
		Environment:   pick(base.Environment, over.Environment),
		Team:          pick(base.Team, over.Team),
		Service:       pick(base.Service, over.Service),
		CustomerID:    pick(base.CustomerID, over.CustomerID),
		UserID:        pick(base.UserID, over.UserID),
		TraceID:       pick(base.TraceID, over.TraceID),
		Feature:       pick(base.Feature, over.Feature),
		PromptKey:     pick(base.PromptKey, over.PromptKey),
		PromptVersion: pick(base.PromptVersion, over.PromptVersion),
	}
	if len(base.Tags) > 0 || len(over.Tags) > 0 {
		out.Tags = make(map[string]string, len(base.Tags)+len(over.Tags))
		for k, v := range base.Tags {
			out.Tags[k] = v
		}
		for k, v := range over.Tags {
			out.Tags[k] = v
		}
	}
	return out
}
